pkg/auth/auth-result.go

```diff
@@ -0,0 +1,71 @@
+package auth
+
+import (
+	"net/http"
+)
+
+type resultType int
+
+const (
+	unknown resultType = iota
+	forbidden
+	unauthorized
+	authorized
+)
+
+type header struct {
+	key   string
+	value string
+}
+
+func Header(key string, value string) header {
+	return header{key, value}
+}
+
+type AuthResult struct {
+	resultType resultType
+	headers    []header
+}
+
+func (a *AuthResult) WriteResponse(w http.ResponseWriter) bool {
+	switch a.resultType {
+	case authorized:
+		// Do not write a response, we're authorized
+		return false
+	case unauthorized:
+		for _, h := range a.headers {
+			w.Header().Add(h.key, h.value)
+		}
+		w.WriteHeader(http.StatusUnauthorized)
+		return true
+	case forbidden:
+		// Return a 404 status even though the issue is that the user is
+		// forbidden so we don't indirectly reveal which repositories are
+		// configured in the bundle server.
+		w.WriteHeader(http.StatusNotFound)
+		return true
```

Comment:
The middleware returns one of the valid response types (currently `Authorized` -> 200, `Unauthorized` (which can have configured headers) -> 401, and `Forbidden` -> 404). So, the middleware chooses the response, but the bundle server decides what to do with that.

Generally speaking, I took a _very_ restrictive approach with this design in an attempt to keep users from writing a plugin that unintentionally reveals too much information; that might be more frustrating than helpful, though. I could instead allow middleware to specify _any_ 4XX code (with optional headers). Then, there's basically two `AuthResult` states: 

1. User is authorized, serve the content
2. User is not authorized, send the specified 4XX error & headers

Does that sound reasonable, or do you think it should be expanded further?

Reply:
> * User is authorized, serve the content
> * User is not authorized, send the specified 4XX error & headers

These seem reasonable, yes.

A middleware author may reasonably want to 200, 401, 403 or 404 with headers for auth.

By the way, technical auth middleware may also want to add headers in the case of a 200 also, for example cookies, DPoP nonces, or [`Authentication-Info`](https://datatracker.ietf.org/doc/html/rfc7615#section-3).

Reply:
Won't Git drop headers it receives from a 200 response? That's the main reason I didn't include a `headers` arg in `Accept()`.

Reply:
> Won't Git drop headers it receives from a 200 response?

Today, yes 😉 

I highly suspect a future where 200s need headers too (passed to credential helpers as part of `store` requests). When Microsoft first tried to push us to access token binding it would require reading headers on 200 and using that data to mint tokens in subsequent requests.

DPoP (draft spec, admittedly) requires headers on 200:
https://datatracker.ietf.org/doc/html/draft-ietf-oauth-dpop#section-8.2

Reply:
Added `headers` to (the once-again-renamed) `Allow()`!